Multithreaded cufft memory leak

### Description

The cufft plan cache does not appear to deallocate GPU memory during thread clean up (GC).  If I launch `cp.fft.fft` on a thread and then exit/join the thread, there is residual memory still allocated on the GPU.


### To Reproduce

```py
from concurrent.futures import ThreadPoolExecutor
import cupy as cp

def task():
    data = cp.ones(2**20)
    return cp.fft.fft(data)

while True:
    with ThreadPoolExecutor(max_workers=3) as executor:
        tasks = [executor.submit(task) for _ in range(10)]
        
    cp.get_default_memory_pool().free_all_blocks()
    cp.get_default_pinned_memory_pool().free_all_blocks()
    cp.fft.config.get_plan_cache().clear()
    
    input("Check nvidia-smi memory usage... Press any key to run again")

```


### Installation

Wheel (`pip install cupy-***`)

### Environment

```
OS                           : Linux-3.10.0-1160.36.2.el7.x86_64-x86_64-with-glibc2.29
Python Version               : 3.8.10
CuPy Version                 : 10.0.0
CuPy Platform                : NVIDIA CUDA
NumPy Version                : 1.21.5
SciPy Version                : 1.7.3
Cython Build Version         : 0.29.24
Cython Runtime Version       : None
CUDA Root                    : /usr/local/cuda
nvcc PATH                    : None
CUDA Build Version           : 11040
CUDA Driver Version          : 11040
CUDA Runtime Version         : 11040
cuBLAS Version               : (available)
cuFFT Version                : 10502
cuRAND Version               : 10205
cuSOLVER Version             : (11, 2, 0)
cuSPARSE Version             : (available)
NVRTC Version                : (11, 4)
Thrust Version               : 101201
CUB Build Version            : 101201
Jitify Build Version         : 60e9e72
cuDNN Build Version          : (not loaded; try `import cupy.cuda.cudnn` first)
cuDNN Version                : (not loaded; try `import cupy.cuda.cudnn` first)
NCCL Build Version           : 21104
NCCL Runtime Version         : 21104
cuTENSOR Version             : None
cuSPARSELt Build Version     : None
Device 0 Name                : Quadro RTX 5000
Device 0 Compute Capability  : 75
Device 0 PCI Bus ID          : 0000:01:00.0

```


### Additional Information

Runtime is within docker image `nvidia/cuda:11.4.2-runtime-ubuntu20.04`

The logs are very long, but the first error is:

```
cupy_backends/cuda/libs/cusparselt.cpp:1469:22: error: 'cusparseLtMatDescAttribute_t' does not name a type; did you mean 'cusparseLtMatmulAlgAttribute_t'?
```
I started with the cupy-feedstock and swapped the source to:

```yaml
source:
  git_url: https://github.com/leofang/cupy.git
  git_tag: fix_fft_cache_leak3
  git_depth: 1 # (Defaults to -1/not shallow)
```

--- a/cupy/fft/_cache.pyx
+++ b/cupy/fft/_cache.pyx
@@ -274,6 +274,9 @@ cdef class PlanCache:
         self._reset()
         self.dev = dev if dev != -1 else runtime.getDevice()
 
+    def __dealloc__(self):
+        self._cleanup()
+
     def __getitem__(self, tuple key):
         # no-op if cache is disabled
         if not self.is_enabled:
@@ -281,8 +284,6 @@ cdef class PlanCache:
             return
 
         cdef _Node node
-        cdef int dev
-        cdef PlanCache cache
         cdef list gpus
 
         node = self.cache.get(key)

